File: java/src/main/java/fr/camory/YatzyThrow.java
package fr.camory;

import io.vavr.collection.List;
import io.vavr.collection.Map;

import static java.util.function.Function.identity;

public class YatzyThrow {
    private final List<Die> dice;

    public YatzyThrow(Die d1, Die d2, Die d3, Die d4, Die d5) {
        dice = List.of(d1, d2, d3, d4, d5);
    }

    List<Die> dice() {
        return dice;
    }

    Map<Die, Integer> diceCount() {
        return dice.groupBy(identity()).mapValues(List::size);
    }

    boolean contains(Die die) {
        return dice.contains(die);
    }
}

File: javascript/lib/yatzy_throw.js
const _ = require('lodash');

class YatzyThrow {
    constructor(d1, d2, d3, d4, d5) {
        [d1, d2, d3, d4, d5].forEach(validate);
        this._dice = [d1, d2, d3, d4, d5];
    }

    get dice() {
        return this._dice;
    }

    get diceCount() {
        return _.mapValues(_.groupBy(this.dice), a => a.length);
    }

    count(die) {
        validate(die);
        return this._dice.indexOf(die) === -1 ? 0 : this.diceCount[die] * die
    }
}

let validate = (die) => {
    if (!_.isInteger(die) || !_.inRange(die, 1, 7)) throw die + " is not a valid value";
};

module.exports = YatzyThrow;
